Align maximally nested DOM timers to 4 milliseconds

rreno opened the pull request:
#### c0b0499d32c9520d52b49e9742b003e61a596e7c
<pre>
Align maximally nested DOM timers to 4 milliseconds
<a href="https://bugs.webkit.org/show_bug.cgi?id=300530">https://bugs.webkit.org/show_bug.cgi?id=300530</a>
<a href="https://rdar.apple.com/160531755">rdar://160531755</a>

Reviewed by Ryosuke Niwa.

Aligning maximally nested DOM timers to a non-zero alignment can prevent abandoned repeating timers
or a large volume of timers from keeping the main thread busy indefinitely. It is especially bad if these
timers end up scheduling more timers which can cause an explosion in outstanding timer-driven work. In the
worst case, the 16ms duty cycle limit for firing timers will be hit but we have so much work backed up we only
sleep the main thread for a few tens of microseconds at most. This keeps CPU perf states
elevated and drives thermal and power use issues.

As a mitigation, we can align DOM timers which have reached their maximum nesting level to
4ms. We already clamp their timeouts to a 4ms minimum but this change will additionally align timeouts to
the nearest 4ms multiple (+/- a randomized proportion, see ScriptExecutionContext::alignedFireTime).

This mitigation is especially helpful on sites with many nested DOM timers or on sites which inadvertantly
leak their DOM timer handles and abandon repeating timers to fire indefinitely. By aligning
these timers to a 4ms multiple we end up lessening timer pile up back pressure and we can sleep the
main thread for at least a few milliseconds at a time between rendering frames. This prevents the main thread
from being more or less continuously scheduled and keeps the system from putting the CPU into maximum perf states
for prolonged or even indefinite periods of time which can drive thermals and draw excessive power.

This tested neutral on our benchmarks with the exception of PLT on iPhone where it is a 4% progression.

* LayoutTests/fast/dom/setInterval-aligned-when-nested-expected.txt: Added.
* LayoutTests/fast/dom/setInterval-aligned-when-nested.html: Added.
* LayoutTests/fast/dom/setTimeout-aligned-when-nested-expected.txt: Added.
* LayoutTests/fast/dom/setTimeout-aligned-when-nested.html: Added.
    Internals tests that check alignment happen when the respective DOM timer reaches its
    maximum nesting level. These are internals tests since trying to observe alignment
    from the web is extremely flaky. It&apos;s easier to tell if a zero-delay timer has been
    clamped, but smaller alignments for longer timeouts are very difficult to observe
    reliably.

* Source/WTF/wtf/Seconds.h:
(WTF::operator*):
(WTF::operator/):
    These should be constexpr.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::domTimerAlignmentInterval const):
    This is where the 4ms alignment is applied when the timer
    has reached its max nesting level.

* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::ScriptExecutionContext::alignedFireTime const):
* Source/WebCore/page/DOMTimer.h:
    The DOM timer interval defaults should be constexpr.

* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::setNextFireTime):
* Source/WebCore/platform/Timer.h:
    Change the TimerAlignment interface from returning an optional fire time
    to instead return a fire time directly, even if no alignment was applied.
    There isn&apos;t a need to differentiate between aligned and unaligned fire times
    being returned from this interface due to testing changes.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isTimerThrottled):
    I adjusted this to account for the 4ms adjustment we will make
    when a timer reaches the max nesting level. I&apos;m not sure if we want to
    call this throttled or not. It is worthwhile to distinguish between
    throttled (like 1s for background timers, for instance) and aligned. The
    former is obviously throttling but the latter is more of a slight adjustment
    to timeouts in order to ensure the main thread is more efficiently scheduled.
(WebCore::Internals::isTimerAligned):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/302278@main">https://commits.webkit.org/302278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fb2f97287b1db7bd9e27f1da3b08cf276a2ded4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39366 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135927 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79964 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/72fbf657-2fcc-4e3e-81e2-fcd208419402) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130407 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/747 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/676 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97851 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65772 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/485f5845-f327-49ea-8ebe-9b6e83e19f47) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131483 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/550 "Found 6 new test failures: fast/events/ios/do-not-show-keyboard-when-focusing-after-blur.html fast/forms/ios/accessory-bar-navigation.html fast/forms/ios/focus-input-via-button.html fast/forms/ios/focus-long-textarea.html fast/forms/ios/zoom-after-input-tap-wide-input.html fast/forms/ios/zoom-after-input-tap.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115167 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78468 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bfb8f8aa-7d98-4eab-b219-5becbd461cba) 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/127886 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/493 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33273 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79210 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120549 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33756 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138376 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126995 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106387 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/682 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111506 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106205 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27061 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/538 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30042 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52975 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/694 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63908 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160017 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/574 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39964 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/635 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/648 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1582,6 +1582,13 @@ int Internals::timerNestingLevel()
     return scriptExecutionContext()->timerNestingLevel();
 }
 
+static bool isTimerAlignedInternal(DOMTimer* timer, const ScriptExecutionContext& context)
```
This function name is rather confusing but I'm not sure what will be a better name.

rreno replied:
Maybe I should just get rid of it. The reason for factoring this logic out was to early return from `isTimerThrottled` but that's not really needed, I don't think.

rreno replied:
Deleted.

rreno committed:
Align maximally nested DOM timers to 4 milliseconds
https://bugs.webkit.org/show_bug.cgi?id=300530
rdar://160531755

Reviewed by Ryosuke Niwa.

Aligning maximally nested DOM timers to a non-zero alignment can prevent abandoned repeating timers
or a large volume of timers from keeping the main thread busy indefinitely. It is especially bad if these
timers end up scheduling more timers which can cause an explosion in outstanding timer-driven work. In the
worst case, the 16ms duty cycle limit for firing timers will be hit but we have so much work backed up we only
sleep the main thread for a few tens of microseconds at most. This keeps CPU perf states
elevated and drives thermal and power use issues.

As a mitigation, we can align DOM timers which have reached their maximum nesting level to
4ms. We already clamp their timeouts to a 4ms minimum but this change will additionally align timeouts to
the nearest 4ms multiple (+/- a randomized proportion, see ScriptExecutionContext::alignedFireTime).

This mitigation is especially helpful on sites with many nested DOM timers or on sites which inadvertantly
leak their DOM timer handles and abandon repeating timers to fire indefinitely. By aligning
these timers to a 4ms multiple we end up lessening timer pile up back pressure and we can sleep the
main thread for at least a few milliseconds at a time between rendering frames. This prevents the main thread
from being more or less continuously scheduled and keeps the system from putting the CPU into maximum perf states
for prolonged or even indefinite periods of time which can drive thermals and draw excessive power.

This tested neutral on our benchmarks with the exception of PLT on iPhone where it is a 4% progression.

* LayoutTests/fast/dom/setInterval-aligned-when-nested-expected.txt: Added.
* LayoutTests/fast/dom/setInterval-aligned-when-nested.html: Added.
* LayoutTests/fast/dom/setTimeout-aligned-when-nested-expected.txt: Added.
* LayoutTests/fast/dom/setTimeout-aligned-when-nested.html: Added.
    Internals tests that check alignment happen when the respective DOM timer reaches its
    maximum nesting level. These are internals tests since trying to observe alignment
    from the web is extremely flaky. It's easier to tell if a zero-delay timer has been
    clamped, but smaller alignments for longer timeouts are very difficult to observe
    reliably.

* Source/WTF/wtf/Seconds.h:
(WTF::operator*):
(WTF::operator/):
    These should be constexpr.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::domTimerAlignmentInterval const):
    This is where the 4ms alignment is applied when the timer
    has reached its max nesting level.

* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::ScriptExecutionContext::alignedFireTime const):
* Source/WebCore/page/DOMTimer.h:
    The DOM timer interval defaults should be constexpr.

* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::setNextFireTime):
* Source/WebCore/platform/Timer.h:
    Change the TimerAlignment interface from returning an optional fire time
    to instead return a fire time directly, even if no alignment was applied.
    There isn't a need to differentiate between aligned and unaligned fire times
    being returned from this interface due to testing changes.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isTimerThrottled):
    I adjusted this to account for the 4ms adjustment we will make
    when a timer reaches the max nesting level. I'm not sure if we want to
    call this throttled or not. It is worthwhile to distinguish between
    throttled (like 1s for background timers, for instance) and aligned. The
    former is obviously throttling but the latter is more of a slight adjustment
    to timeouts in order to ensure the main thread is more efficiently scheduled.
(WebCore::Internals::isTimerAligned):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/302278@main
